sms_toolkit/utils/encoding.py

```diff
@@ -26,10 +31,20 @@ def convert_to_unicode_characters(string):
 
 
 def encode_unicode_character_to_utf16(character):
-    if len(character) > 1:
-        return flatten(map(encode_unicode_character_to_utf16, character))
-
-    return [((0xff00 & ord(character)) >> 8), 0x00ff & ord(character)]
+    """
+    Encode the given unicode code points into utf-16 and return a list a of bytes. Note that for code-points within
+    the BMP the number of bytes would be 2, but for code-points not in BMP, the number of bytes will be 4.
+
+    Args:
+        character: A unicode code-point
+
+    Returns:
+        A list of the bytes
+    """
+    if is_python_3():
+        return [c for c in character.encode('utf-16-be')]
+    else:
+        return [ord(c) for c in character.encode('utf-16-be')]
```

Comment:
Why is this change required? 
The previous approach had one major flaw - using `len`. Depending on whether the python interpreter build is narrow or wide (or in < py3.3 where the string encoding is based off of the highest code-point), the behaviour of len changes if the input is a code-point that's greater than what can be fit in 2 bytes (UTF-16 encodes them in 2 2-byte pairs, which we need to detect, since UCS-2 has a fixed 2 bytes limit).

In the narrow build - py2.7.12 (u'\U0001f600' is the unicode code point for 😀) - 
``` python 
>>> len(u'\U0001f600')
2
```

In wide build - py 2.7.12 
``` python 
>>> len(u'\U0001f600')
1
```

In python 3.5 (No need for u string prefix)- 
``` python 
>>> len('\U0001f600')
1
```
The issue here is that we want the len to be 2 everywhere. So we need to encode it to utf-16 explicitly.

Why the `be` in `utf-16-be` though?
Encodings can be done using big endian or little endian. If we don't explicitly state that, then will be 2 bytes which will be used to explicitly say what endianess the compiler has used. That information is not required for us. We can directly tell the interpreter to use big-endian when encoding.

Why do we need to handle py2 and py3 separately?
str.encode behaves differently in python 2 and python 3.
In python 2 encode returns an 8-bit string version of the Unicode string, encoded in the requested encoding ([Source](https://docs.python.org/2/howto/unicode.html)).
In python 3, encode returns an encoded version of the string as a bytes object. ([Source](https://docs.python.org/3/library/stdtypes.html#str.encode))

We basically want the list of byte values for the given character here. In python 3, we directly get it by iterating through every entry of encode. In py2, since the value is a string version, we can get the value by `ord`ing it.